server.go

```diff
@@ -631,13 +634,36 @@ func (srv *Server) disposeBuffer(w *response) {
 
 func (srv *Server) serveDNS(w *response) {
 	req := new(Msg)
-	err := req.Unpack(w.msg)
-	if err != nil { // Send a FormatError back
-		x := new(Msg)
-		x.SetRcodeFormatError(req)
-		w.WriteMsg(x)
+	dh, off, err := unpackMsgHdr(w.msg, 0)
+	req.setHdr(dh)
+
+	if err != nil {
+		// Let client hang, they are sending crap; any reply can be used to amplify.
+		return
+	}
+
+	// Response MUST be handled separately, because we can't reply to these messages, if we
+	// do we can ping-ping ourselves to death.
+	response := dh.Bits&_QR != 0
+	if response {
+		return
+	}
+
```

Comment:
```go
req := new(Msg)
req.setHdr(dh)
```

should be moved right here. `req` isn't used before now.

Reply:
good catch. done.